### PR TITLE
Supplychain refactor

### PR DESCRIPTION
**Pull request by cianx**
Addresses the following Jira tasks:
STL-270 - Supplychain Transaction Processor Spec
STL-271 - Update the Supplychain TP to handle Hash collisions. 
STL-272 - Update the Supplychain TP to use ProtoBuf for global state representation
STL-316 - Update Supply chain records to use native addressing
STL-317 - Refactor the Application to allow it to be tracked

### CODE REVIEW

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+developed. As such it has not been thoughly tested and there will be
+protentially breaking changes made to it.
+
+State
+=====
+
+This section describes in detail how SCTP objects are stored and addressed in
+the global state. All objects in state are stored as encoded protobufs. Each
+type of object below is stored in containers at nodes in the Merkle Trie. The
+address of the Merkle nodes is detailed below in the addressing section. The entry at the nodes is a container object that allows for multiple items to be stored at that node if there is an address collision.
```
Comment (jsmitchell) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: Need to describe the storage format for collisions (in the state section), lexicographically sorted by natural keys, etc.

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+    agent_addr = namespace + sha512(agent_pub_key)).hexdigest()
+
+Application addresses are generated as the concatination of the hash(sha256)
+applicant agent public key and the hash(sha256) record_identifier. sha256 is
+used here to so the top half of the address can identify the Agent and the
+bottom half can identify the Record.
+
+.. code-block:: javascript
+
+    agent_addr = namespace + sha256(agent_pub_key)).hexdigest()
```
Comment (jsmitchell) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: How do you find open applications for a Record? It looks like any Agent can submit applications, so these aren't forward derivable in an efficient way without an index.

Comment (jsmitchell) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: If this were limited to current custodian/owner, then it would be derivable.

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+
+
+Agent Transactions
+------------------
+
+Create Agent
+++++++++++++
+
+Create an agent record, agent address will be calculated from the public key
+extracted from the transactions signature.
```
Comment (jsmitchell) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: Public key doesn't need to be recovered. It's on the header.

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+Accept Application
+++++++++++++++++++
+
+Accept an application for transfer of ownership or custodianship of the record. Must be submitted by the current owner or custodian.
+
+.. code-block:: javascript
+
+    message ApplicationAcceptPayload {
+        string record_identifier = 1;
+        string applicant = 2;
```
Comment (jsmitchell) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: Is this supposed to be the pub key of the agent who created the application?

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+
+Execution
+=========
+
+Agents can only be created by transactions signed with their private key.
+
+Records exist in a active modifiable state until they are finalized. Once a
+Record is finalized it can be thought of as the item it represents is
+destroyed. Either consumed in manufacturing, lost, or the victim of an
```
Comment (jsmitchell) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: Sentence fragment

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+unfortunate accident.
+
+Application can only be created by the applicant agent and then transition to
```
Comment (jsmitchell) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: s/Application/Applications/

Comment (jsmitchell) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: s/transition/transitioned/

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+either an Accepted state which means the proposal has been excuted and the
```
Comment (jsmitchell) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: s/excuted/executed/

File: bin/protogen [hunk 8]
```diff
@@ -62,6 +62,12 @@ def main(args=sys.argv[1:]):
             "consensus/poet/families/sawtooth_validator_registry/protos")
     protoc(proto_dir, "consensus/poet/common", "sawtooth_poet_common/protobuf")
 
+    # 4. Generate supplychain protos
+    proto_dir = JOIN(TOP_DIR, "families/supplychain_python/protos")
```
Comment (jsmitchell) [inline bin/protogen]: It feels like the family name shouldn't contain the implementation language. The python specific implementation should be one level lower, and then the protos can be at the level that is language agnostic.

File: families/supplychain_python/sawtooth_supplychain/processor/handler.py [hunk 9]
```diff
@@ -14,442 +14,410 @@
...
+
+    def _agent_create(self, state, originator, data):
+        txn_data = AgentCreatePayload()
+        txn_data.ParseFromString(data)
+
+        agent_addr = Addressing.agent_address(originator)
+        state_items = self._get(state, [agent_addr])
+        agents = state_items.get(agent_addr, AgentContainer())
+
+        # check that the agent does not exists
```
Comment (jsmitchell) [inline families/supplychain_python/sawtooth_supplychain/processor/handler.py]: s/exists/exist/

File: families/supplychain_python/sawtooth_supplychain/processor/handler.py [hunk 10]
```diff
@@ -14,442 +14,410 @@
...
+        state_items = self._get(state, [agent_addr, record_addr])
+
+        agents = state_items.get(agent_addr, None)
+        records = state_items.get(record_addr, RecordContainer())
+
+        # check that the agent exists
+        if self._find_agent(agents, originator) is None:
+            raise InvalidTransaction("Agent is not registered.")
+
+        # check that the record does not exists
```
Comment (jsmitchell) [inline families/supplychain_python/sawtooth_supplychain/processor/handler.py]: s/exists/exist/

File: families/supplychain_python/sawtooth_supplychain/processor/handler.py [hunk 11]
```diff
@@ -14,442 +14,410 @@
...
+        txn_data = RecordFinalizePayload()
+        txn_data.ParseFromString(data)
+
+        record_addr = Addressing.record_address(txn_data.identifier)
+
+        state_items = self._get(state, [record_addr])
+
+        records = state_items.get(record_addr, RecordContainer())
+
+        # check that the record does not exists
```
Comment (jsmitchell) [inline families/supplychain_python/sawtooth_supplychain/processor/handler.py]: s/exists/exist/

File: families/supplychain_python/sawtooth_supplychain/processor/handler.py [hunk 12]
```diff
@@ -14,442 +14,410 @@
...
+        record = self._find_record(records, txn_data.identifier)
+        if record is None:
+            raise InvalidTransaction("Record does not exists.")
```
Comment (jsmitchell) [inline families/supplychain_python/sawtooth_supplychain/processor/handler.py]: s/exists/exist/

File: families/supplychain_python/sawtooth_supplychain/processor/handler.py [hunk 13]
```diff
@@ -14,442 +14,410 @@
...
+        records = state_items.get(record_addr, None)
+        record = self._find_record(records, txn_data.record_identifier)
+        if record is None:
+            raise InvalidTransaction("Record does not exist.")
+
+        if record.final:
+            raise InvalidTransaction("Record is final, no updates can be " +
+                                     "proposed.")
+
+        # check that the application does not exists
```
Comment (jsmitchell) [inline families/supplychain_python/sawtooth_supplychain/processor/handler.py]: s/exists/exist/

File: families/supplychain_python/sawtooth_supplychain/processor/handler.py [hunk 14]
```diff
@@ -14,442 +14,410 @@
...
+            txn_data.record_identifier)
+
+        state_items = self._get(state, [application_addr, record_addr])
+
+        # check that the application exists
+        applications = state_items.get(application_addr, None)
+        application = self._find_application(
+            applications, txn_data.applicant, txn_data.record_identifier)
+        if application is None:
+            raise InvalidTransaction("Application does not exists.")
```
Comment (jsmitchell) [inline families/supplychain_python/sawtooth_supplychain/processor/handler.py]: s/exists/exist/

File: families/supplychain_python/sawtooth_supplychain/processor/handler.py [hunk 15]
```diff
@@ -14,442 +14,410 @@
...
+
+        # check that the recorf exists
```
Comment (jsmitchell) [inline families/supplychain_python/sawtooth_supplychain/processor/handler.py]: s/recorf/record/

File: families/supplychain_python/sawtooth_supplychain/processor/handler.py [hunk 16]
```diff
@@ -14,442 +14,410 @@
...
+        records = state_items.get(record_addr, None)
+        record = self._find_record(records, application.record_identifier)
+        if record is None:
+            raise InvalidTransaction("Record does not exists.")
```
Comment (jsmitchell) [inline families/supplychain_python/sawtooth_supplychain/processor/handler.py]: s/exists/exist/

File: families/supplychain_python/sawtooth_supplychain/processor/handler.py [hunk 17]
```diff
@@ -14,442 +14,410 @@
...
+            txn_data.record_identifier)
+
+        state_items = self._get(state, [application_addr])
+
+        # check that the application exists
+        applications = state_items.get(application_addr, None)
+        application = self._find_application(
+            applications, txn_data.applicant, txn_data.record_identifier)
+        if application is None:
+            raise InvalidTransaction("Application does not exists.")
```
Comment (jsmitchell) [inline families/supplychain_python/sawtooth_supplychain/processor/handler.py]: s/exists/exist/

File: families/supplychain_python/sawtooth_supplychain/processor/handler.py [hunk 18]
```diff
@@ -14,442 +14,410 @@
...
+            txn_data.record_identifier)
+
+        state_items = self._get(state, [application_addr, record_addr])
+
+        # check that the application exists
+        applications = state_items.get(application_addr, None)
+        application = self._find_application(
+            applications, txn_data.applicant, txn_data.record_identifier)
+        if application is None:
+            raise InvalidTransaction("Application does not exists.")
```
Comment (jsmitchell) [inline families/supplychain_python/sawtooth_supplychain/processor/handler.py]: s/exists/exist/

File: families/supplychain_python/sawtooth_supplychain/processor/handler.py [hunk 19]
```diff
@@ -14,442 +14,410 @@
...
+
+        # check that the recorf exists
```
Comment (jsmitchell) [inline families/supplychain_python/sawtooth_supplychain/processor/handler.py]: s/recorf/record/

File: families/supplychain_python/sawtooth_supplychain/processor/handler.py [hunk 20]
```diff
@@ -14,442 +14,410 @@
...
+        records = state_items.get(record_addr, None)
+        record = self._find_record(records, application.record_identifier)
+        if record is None:
+            raise InvalidTransaction("Record does not exists.")
```
Comment (jsmitchell) [inline families/supplychain_python/sawtooth_supplychain/processor/handler.py]: s/exists/exist/

File: families/supplychain_python/sawtooth_supplychain/processor/handler.py [hunk 21]
```diff
@@ -14,442 +14,410 @@
...
-        AgentHandler.update_record_tracking(
-            state, identifier,
-            [("AcceptedApplications", record_id, sensor_id, False)])
-
-    @classmethod
-    def remove_accepted_application(cls, state, identifier, record_id):
-        AgentHandler.update_record_tracking(
-            state, identifier,
-            [("AcceptedApplications", record_id, "del", False)])
+            raise InternalError("Error setting state nothing updated?")
```
Comment (jsmitchell) [inline families/supplychain_python/sawtooth_supplychain/processor/handler.py]: I think this should be an InvalidTransaction rather than an InternalError.

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 22]
```diff
@@ -0,0 +1,317 @@
+***************************************
```
Comment (peterschwarz) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: Split this out to a separate PR, so the code can be validated against, once all the changes have been made to the document.

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 23]
```diff
@@ -0,0 +1,317 @@
...
+
+The Supplychain Transaction Processor(SCTP) provides a mechanism for
+storing records of asset ownership. The SCTP tracks records, which correlate
+to a real world item. Every Record has an owner and a custodaian, which can
+be the same Agent. The owner is the Agent the legally owns the good. The
+custodian is the Agent that is in possesion of the good, which may be a
+warehousing service, transportation service, etc. Agents are actors in the
+system that are can manupulate the Records by updating them or transfering
+their role to other Agents. Agents are identified by their public key and
+Authenticated by the signature on the transactions they submit. Transfer of
```
Comment (peterschwarz) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: Authenticated should not be capitalized.

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 24]
```diff
@@ -0,0 +1,317 @@
...
+roles(Owner, Custodian) betweeen Agents is done via an Application process,
```
Comment (peterschwarz) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: space before `(`

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 25]
```diff
@@ -0,0 +1,317 @@
+***************************************
+Supply Chain Transaction Family
+***************************************
+
+Overview
+=========
+
+The Supplychain Transaction Processor(SCTP) provides a mechanism for
```
Comment (peterschwarz) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: Space before `(`

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 26]
```diff
@@ -0,0 +1,317 @@
...
+The strings are converted to a Merkle prefix by taking the first 6 characters
+of the hex encoded sha512 hash of the string. .
+
+Agent addresses are generated sha512(agent_pub_key), prepended with the supplychain.agent namespace.
+
+.. code-block:: javascript
+
+    agent_addr = namespace + sha512(agent_pub_key)).hexdigest()
+
+Application addresses are generated as the concatination of the hash(sha256)
```
Comment (peterschwarz) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: Why do these use SHA256, and the other uses SHA 512?

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 27]
```diff
@@ -0,0 +1,317 @@
...
+    {
+        family_name: "sawtooth_supplychain"
+        family_version: "0.5"
+        encoding: "application/protobuf"
+    }
+
+Inputs and Outputs
+++++++++++++++++++
+The inputs for config family transactions must include:
+the address of the object the transaction is operating on
```
Comment (peterschwarz) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: Probably should be a bulleted list

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 28]
```diff
@@ -0,0 +1,317 @@
...
+
+    {
+        family_name: "sawtooth_supplychain"
+        family_version: "0.5"
+        encoding: "application/protobuf"
+    }
+
+Inputs and Outputs
+++++++++++++++++++
+The inputs for config family transactions must include:
```
Comment (peterschwarz) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: This is not the config family.

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 29]
```diff
@@ -0,0 +1,317 @@
...
+the address of the object the transaction is operating on
+if it is a record operation it must contain any Agents involved in the
+operation.
+The outputs for config family transactions must include:
```
Comment (peterschwarz) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: Still not the config family

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst (hunk 22 above)
Comment (cianx) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: @peterschwarz I don't understand the reasoning on this. The code and spec are both updated to match. It seems that you are requesting that I create a situation where I have the spec and code out of sync.

File: families/supplychain/patch1 [hunk 30]
```diff
@@ -0,0 +1,202 @@
+From 5976b4b2808e0aa3eb6aa65244aa9cf04bf78f37 Mon Sep 17 00:00:00 2001
```
Comment (jsmitchell) [inline families/supplychain/patch1]: Whoops. Do you want this in here?

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 31]
```diff
@@ -0,0 +1,241 @@
...
+        family_name: "sawtooth_supplychain"
+        family_version: "0.5"
+        encoding: "application/protobuf"
+    }
+
+Inputs and Outputs
+++++++++++++++++++
+The inputs for Supplychain family transactions must include:
+- the address of the object the transaction is operating on if it is a record operation,
+- Any Agents involved in theoperation.
```
Comment (jsmitchell) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: s/theoperation/the operation/

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 32]
```diff
@@ -0,0 +1,241 @@
...
+
+The outputs for Supplychain family transactions must include:
+- The address of the object the transaction is operating on if it a Record operation
```
Comment (jsmitchell) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: s/if it a/if it is a/

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 33]
```diff
@@ -0,0 +1,241 @@
...
+- Any Agents involved in the operation.
+
+Dependencies
+++++++++++++
+
+Transactions should be batch and specify any transactions that they expect
```
Comment (jsmitchell) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: s/batch/batched/ ?

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst [hunk 34]
```diff
@@ -0,0 +1,241 @@
...
+to be commited as dependencies.
+
+
+Transaction Payload
+===================
+
+All SCTP transactions are wrapped in a payload object that allows for the items to be dispatch to the correct handling logic.
```
Comment (jsmitchell) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: 80 column wrap

File: docs/source/transaction_family_specifications/supplychain_transaction_family.rst (hunk 22 above)
Comment (peterschwarz) [inline docs/source/transaction_family_specifications/supplychain_transaction_family.rst]: Specs are design docs.  We're trying to promote the proper way to develop transaction families: Write a spec, iterate on it, then write the transaction processor.  By doing this simultaneously, it's difficult to know which is actually correct: the spec or the code.
